### ISSUE TITLE
How any epochs to be run in Bigdl.sh

### ISSUE THREAD

**Issue by jaymahesh** [intel-analytics/BigDL#464]
I have 64 cores, 1 node and 32g driver memory allocated. I am running the training, but it has already crossed 12 epcochs. Wanted to check the number of epoch and time that this bigdl.sh will run in such a configuration.

**Comment by jason-dai** [comment]
The number of epoch is specified in the user code; which example are you running?

**Comment by jaymahesh** [comment]
I downloaded the whole zip file from https://github.com/intel-analytics/BigDL

May I know which 'user code' are you refering to? Thanks

**Comment by jason-dai** [comment]
There are many different models provided in BigDL (e.g., vgg, resnet, rnn, etc.). What's your command to launch the training program?

**Comment by jaymahesh** [comment]
it is vgg

**Comment by jaymahesh** [comment]
there are some 42 classes in VGG folder, so how many epochs would be run? Any feedback would be helpful, thanks

**Comment by jaymahesh** [comment]
My command to launch the program is:
spark-sumbit --class com.intel.analytics.bigdl.models.vgg.Train --driver-memory 32g \  /home/jaym/BigDL-master/dist/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies.jar \ 
--core 64 \
--node 1 \
--env spark \
-f /home/jaym/cifar-10-batches-bin/ \
-b 256

**Comment by jason-dai** [comment]
The default epoch is 90 for vgg; you may specify a different value using --maxEpoch (or -e).

**Comment by jaymahesh** [comment]
Hi Jason I think it would be --maxEpochS right? By the way, what is the effect of reducing the the Epochs?

Secondly, how do we visualize the train-validate-test the output?

**Comment by jason-dai** [comment]
An epoch means a full pass through the training set. And you can see the validation results in the log.

**Comment by jaymahesh** [comment]
I reduced the maxEpoch and it runs. However I would like to understand 2 things
1) What is the model_file shown below? Where from do we get or load that?
./dist/bin/bigdl.sh -- \
spark-submit --master local[physical_core_number] \
--class com.intel.analytics.bigdl.models.vgg.Test \
dist/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies-all-in-one.jar \
-f cifar \
--model model_file \
2)
Are the logs stored in the same /BigDL-master/dist/bin directory or elsewhere? Is there a way to visualize the output for the VGG data?
Thanks

**Comment by i8run** [comment]
Hi Jaymahesh,

For Q1, the **model_file** is the model have been saved somewhere. So
+   How to generate the model ?

    ```
    ./dist/bin/bigdl.sh -- \
    spark-submit --class com.intel.analytics.bigdl.models.vgg.Train \
    dist/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies.jar \
    --core physical_core_number \
    --node node_number \
    --env spark \
    -f Cifar-folder \
    -b batchsize \
    --checkpoint ~/model
    ```

    This command will generate the model you want under *model* which is controlled by `--checkpoint` .

+   How to load the model ?

    ```
    ./dist/bin/bigdl.sh -- \
    spark-submit --class com.intel.analytics.bigdl.models.vgg.Test \
    dist/lib/bigdl-0.1.0-SNAPSHOT-jar-with-dependencies-all-in-one.jar \
    -f cifar-folder \
    --model model_file \
    --nodeNumber node_number \
    --core physical_core_number \
    --env spark \
    -b batch_size
    ```

    This command will load the model generated by the command above, which is controlled by `--model`.

For Q2, Does the **log** means *the loss of every iteration/epoch* ? If so,
+   Loss info was printed on **stdout** where you submit the job.  You can use Linux command `nohup`, `tee` or some other IO redirection method to save  the log.

+   Other intermediate file is under the path specified by `--checkpoint` .

+   And currently, we don't support visualize the loss/accuracy. Maybe you can use `gnuplot` with a little scripts to visualize it.

**Comment by jaymahesh** [comment]
Hi 
Thanks.
In my question I meant 'validation / test results' by 'log'. 
Are they stored in a specific folder for us to visualize?

Thanks

**Comment by i8run** [comment]
Currently, the validation results are mixed with training info. And all the info/log is printed on the `stdout` where you submit the job.
So you should have to `grep` the output.

Validation results are similar as below, which will be printed every epoch by default.

```
2017-02-15 15:08:25 INFO  DistriOptimizer$:427 - [Wall Clock 558.898389713s] Validate model...
2017-02-15 15:08:27 INFO  DistriOptimizer$:467 - top1 accuracy is Accuracy(correct: 4343, count: 10000, accuracy: 0.4343)
```

Training info is similar as below, which will be printed every iteration.

```
2017-02-15 15:13:16 INFO  DistriOptimizer$:241 - [Epoch 10 28672/50000][Iteration 537][Wall Clock 846.836487799s] Train 896 in 1.33303526seconds. Throughput is 672.1502625519448 records/second. Loss is 1.216449067794851. 
```

**Comment by jaymahesh** [comment]
I have successfully completed the Train and Test cycles for VGG with maxEpoch 1 and 64 cores. I just wanted to capture the results so that I can visualize. Looks like you are saying I need to do so with grep output and no other form of output is available.

**Comment by jason-dai** [comment]
@jaymahesh I wonder what types of visualizations you have in mind. We are currently working on python support for BigDL, which would allow the user to use notebooks like Jupyter for visualizations; let us know if you have specific requirements on that.

**Comment by jaymahesh** [comment]
Yes, would love to have python and/or R support, if you can provide that, thanks

**Comment by jaymahesh** [comment]
Hi Jason, in Python or R, I would like to to plot the output of the model being 'test'

**Comment by jaymahesh** [comment]
Jason

When I ran the train and test, here is what I get, especially on the accuracy, Pls note for Train I specified **maxEpoch 1**, so would that impact the accuracy?

**Train:**
2017-02-23 10:03:01 INFO  DistriOptimizer$:285 - [Epoch 1 49920/50000][Iteration 196][Wall Clock 1955.301932307s] Epoch finished. Wall clock time is 1963479.67299ms
2017-02-23 10:03:01 INFO  DistriOptimizer$:427 - [Wall Clock 1963.47967299s] Validate model...
2017-02-23 10:04:23 INFO  DistriOptimizer$:467 - top1 accuracy is Accuracy(correct: 991, count: 10000, accuracy: 0.0991)
[Wall Clock 1963.47967299s] Save model to /root/model

**Test:**
2017-02-23 10:15:52 INFO  Remoting:74 - Remoting started; listening on addresses :[akka.tcp://sparkDriverActorSystem@192.168.1.202:52428]
2017-02-23 10:16:04 INFO  DistriValidator$:50 - model thread pool size is 1
top1 accuracy is Accuracy(correct: 1000, count: 10000, accuracy: 0.1)

**Comment by yiheng** [comment]
one epoch is not enough. The default epoch number is 90.

**Comment by jason-dai** [comment]
Closing this issue; if there are further questions, please send them to BigDL user group (https://groups.google.com/forum/#!forum/bigdl-user-group).
